src/p2p/swarm.rs

```diff
@@ -310,71 +324,51 @@ impl NetworkBehaviour for SwarmApi {
-                    .unwrap_or_default()
-                    .into_iter(),
-            );
-
-        for addr in failed {
-            self.connect_registry
-                .finish_subscription(addr.into(), Err("disconnected".into()));
-        }
-    }
-
     fn inject_event(&mut self, _peer_id: PeerId, _connection: ConnectionId, _event: void::Void) {}
 
-    fn inject_dial_failure(&mut self, peer_id: &PeerId) {
-        trace!("inject_dial_failure: {}", peer_id);
-        if self.pending_addresses.contains_key(peer_id) {
-            // it is possible that these addresses have not been tried yet; they will be asked
-            // for soon.
-            self.events
-                .push_back(swarm::NetworkBehaviourAction::DialPeer {
-                    peer_id: *peer_id,
-                    condition: DialPeerCondition::NotDialing,
-                });
-        }
-
-        // this should not be executed once, but probably will be in case unsupported addresses or something
-        // surprising happens.
-        for failed in self.pending_connections.remove(peer_id).unwrap_or_default() {
-            self.connect_registry
-                .finish_subscription(failed.into(), Err("addresses exhausted".into()));
-        }
-    }
-
-    fn inject_addr_reach_failure(
+    fn inject_dial_failure(
         &mut self,
-        peer_id: Option<&PeerId>,
-        addr: &Multiaddr,
-        error: &dyn std::error::Error,
+        peer_id: Option<PeerId>,
+        _handler: Self::ConnectionHandler,
+        error: &DialError,
     ) {
-        trace!("inject_addr_reach_failure {} {}", addr, error);
-
         if let Some(peer_id) = peer_id {
-            match self.pending_connections.entry(*peer_id) {
+            match self.pending_connections.entry(peer_id) {
                 Entry::Occupied(mut oe) => {
                     let addresses = oe.get_mut();
-                    let addr = MultiaddrWithPeerId::try_from(addr.clone())
-                        .expect("dialed address contains peerid in libp2p 0.38");
-                    let pos = addresses.iter().position(|a| *a == addr);
 
-                    if let Some(pos) = pos {
-                        addresses.swap_remove(pos);
-                        self.connect_registry
-                            .finish_subscription(addr.into(), Err(error.to_string()));
+                    match error {
+                        DialError::Transport(multiaddrs) => {
+                            for (addr, error) in multiaddrs {
+                                let addr = MultiaddrWithPeerId::try_from(addr.clone())
+                                    .expect("to recieve an MultiAddrWithPeerId from DialError");
+                                self.connect_registry
+                                    .finish_subscription(addr.into(), Err(error.to_string()));
+                            }
+
+                            let peer_ids = multiaddrs
+                                .into_iter()
```

Comment:
```suggestion
                                .iter()
```

The error is behind a borrow so cannot take ownership here. I was going to ask about the `addr.clone()` on line 352 but this explains it.

However ... This vec building seems extra because we are already handling a failure to dial a specific peer, and we only allow connect's to `MultiaddrWithPeerId` on line 97 with `pub fn connect(&mut self, addr: MultiaddrWithPeerId) -> ...`.

So perhaps it would be better to remove from `addresses` with in the `finish_subscription` loop above. It will need to be position based swap remove.